### PR TITLE
table: Update delta from column slice, instead of map

### LINKED ISSUE #53660: TransactionContext.UpdateDeltaForTable is inefficient
## Enhancement
In a microbenchmark of AddRecord (with pipelined DML, ART memdb and #53605), UpdateDeltaForTable uses noticeable CPU time.
image
The primary time consumption is for map iteration, which could be optimized.
```
ROUTINE ======================== github.com/pingcap/tidb/pkg/sessionctx/variable.(*TransactionContext).UpdateDeltaForTable in /home/eke/code/tidb/pkg/sessionctx/variable/session.go
     180ms      1.38s (flat, cum)  5.90% of Total
         .          .    309:func (tc *TransactionContext) UpdateDeltaForTable(physicalTableID int64, delta int64, count int64, colSize map[int64]int64) {
      10ms       10ms    310:	tc.tdmLock.Lock()
      40ms       40ms    311:	defer tc.tdmLock.Unlock()
         .          .    312:	if tc.TableDeltaMap == nil {
         .          .    313:		tc.TableDeltaMap = make(map[int64]TableDelta)
         .          .    314:	}
      20ms       30ms    315:	item := tc.TableDeltaMap[physicalTableID]
      30ms       30ms    316:	if item.ColSize == nil && colSize != nil {
         .       10ms    317:		item.ColSize = make(map[int64]int64, len(colSize))
         .          .    318:	}
         .          .    319:	item.Delta += delta
         .          .    320:	item.Count += count
         .          .    321:	item.TableID = physicalTableID
      50ms      990ms    322:	for key, val := range colSize {
         .      160ms    323:		item.ColSize[key] += val
         .          .    324:	}
      20ms       90ms    325:	tc.TableDeltaMap[physicalTableID] = item
      10ms       20ms    326:}
         .          .    327:
         .          .    328:// GetKeyInPessimisticLockCache gets a key in pessimistic lock cache.
         .          .    329:func (tc *TransactionContext) GetKeyInPessimisticLockCache(key kv.Key) (val []byte, ok bool) {
         .          .    330:	if tc.pessimisticLockCache == nil && tc.CurrentStmtPessimisticLockCache == nil {
         .          .    331:		return nil, false
```

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: close #53660 

Problem Summary:

`AddRecord`/`UpdateRecord`/`RemoveRecord` is in the hot path for writing. Iterating over a map when they call `UpdateDeltaForTable` is slow. 

### What changed and how does it work?

Replace the map with a slice.

Before:
```
ROUTINE ======================== github.com/pingcap/tidb/pkg/sessionctx/variable.(*TransactionContext).UpdateDeltaForTable in /Users/eke/code/tidb/pkg/sessionctx/variable/session.go
      30ms      420ms (flat, cum)  2.04% of Total
         .          .    309:func (tc *TransactionContext) UpdateDeltaForTable(physicalTableID int64, delta int64, count int64, colSize map[int64]int64) {
         .          .    310:	tc.tdmLock.Lock()
         .          .    311:	defer tc.tdmLock.Unlock()
         .          .    312:	if tc.TableDeltaMap == nil {
         .          .    313:		tc.TableDeltaMap = make(map[int64]TableDelta)
         .          .    314:	}
         .          .    315:	item := tc.TableDeltaMap[physicalTableID]
         .          .    316:	if item.ColSize == nil && colSize != nil {
         .          .    317:		item.ColSize = make(map[int64]int64, len(colSize))
         .          .    318:	}
         .          .    319:	item.Delta += delta
         .          .    320:	item.Count += count
         .          .    321:	item.TableID = physicalTableID
      10ms      350ms    322:	for key, val := range colSize {
         .       20ms    323:		item.ColSize[key] += val
         .          .    324:	}
         .       30ms    325:	tc.TableDeltaMap[physicalTableID] = item
      20ms       20ms    326:}
         .          .    327:
         .          .    328:// GetKeyInPessimisticLockCache gets a key in pessimistic lock cache.
         .          .    329:func (tc *TransactionContext) GetKeyInPessimisticLockCache(key kv.Key) (val []byte, ok bool) {
         .          .    330:	if tc.pessimisticLockCache == nil && tc.CurrentStmtPessimisticLockCache == nil {
         .          .    331:		return nil, false
```

This PR:
```
ROUTINE ======================== github.com/pingcap/tidb/pkg/sessionctx/variable.(*TransactionContext).UpdateDeltaForTableFromColSlice in /Users/eke/code/tidb/pkg/sessionctx/variable/session.go
      10ms       30ms (flat, cum)  0.13% of Total
         .          .    334:func (tc *TransactionContext) UpdateDeltaForTableFromColSlice(
         .          .    335:	physicalTableID int64, delta int64,
         .          .    336:	count int64, colSizes []ColSize,
         .          .    337:) {
         .          .    338:	tc.tdmLock.Lock()
         .          .    339:	defer tc.tdmLock.Unlock()
         .          .    340:	if tc.TableDeltaMap == nil {
         .          .    341:		tc.TableDeltaMap = make(map[int64]TableDelta)
         .          .    342:	}
         .       10ms    343:	item := tc.TableDeltaMap[physicalTableID]
         .          .    344:	if item.ColSize == nil && colSizes != nil {
         .          .    345:		item.ColSize = make(map[int64]int64, len(colSizes))
         .          .    346:	}
         .          .    347:	item.Delta += delta
         .          .    348:	item.Count += count
         .          .    349:	item.TableID = physicalTableID
         .          .    350:	for _, s := range colSizes {
         .       10ms    351:		item.ColSize[s.ColId] += s.Size
         .          .    352:	}
         .          .    353:	tc.TableDeltaMap[physicalTableID] = item
      10ms       10ms    354:}
         .          .    355:
         .          .    356:// GetKeyInPessimisticLockCache gets a key in pessimistic lock cache.
         .          .    357:func (tc *TransactionContext) GetKeyInPessimisticLockCache(key kv.Key) (val []byte, ok bool) {
         .          .    358:	if tc.pessimisticLockCache == nil && tc.CurrentStmtPessimisticLockCache == nil {
         .          .    359:		return nil, false
```

### Check List

Tests 

- [x] Unit test: microbenchmark
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -137,7 +137,7 @@ require (
 	golang.org/x/time v0.5.0
```
Comment (hawkingrei) [inline go.mod:137]: Please run ```make bazel_prepare```.

File: go.mod [hunk 2]
```diff
@@ -137,7 +137,7 @@ require (
...
 	golang.org/x/tools v0.21.0
 	google.golang.org/api v0.162.0
-	google.golang.org/grpc v1.63.2
+	google.golang.org/grpc v1.64.0
```
Comment (cfzjywxk) [inline go.mod:140]: Is this an unrelated change?

File: pkg/sessionctx/variable/session.go [hunk 3]
```diff
@@ -325,6 +325,34 @@ func (tc *TransactionContext) UpdateDeltaForTable(physicalTableID int64, delta i
...
+	physicalTableID int64, delta int64,
+	count int64, colSizes []ColSize,
+) {
+	tc.tdmLock.Lock()
+	defer tc.tdmLock.Unlock()
+	if tc.TableDeltaMap == nil {
+		tc.TableDeltaMap = make(map[int64]TableDelta)
+	}
+	item := tc.TableDeltaMap[physicalTableID]
+	if item.ColSize == nil && colSizes != nil {
```
Comment (cfzjywxk) [inline pkg/sessionctx/variable/session.go:344]: Perhaps `len(colSizes) > 0`  could help to avoid some non-nil zero slice situations compared with `colSizes != nil`.

File: pkg/table/tables/tables.go [hunk 4]
```diff
@@ -670,7 +670,7 @@ func (t *TableCommon) UpdateRecord(ctx context.Context, sctx table.MutateContext
 			return err
 		}
 	}
-	colSize := make(map[int64]int64, len(t.Cols()))
+	colSize := make([]variable.ColSize, len(t.Cols()))
```
Comment (cfzjywxk) [inline pkg/table/tables/tables.go:673]: Nice catch, we could check if there similar implemetations in other performance critical paths.

File: pkg/sessionctx/variable/session.go [hunk 5]
```diff
@@ -325,6 +325,35 @@ func (tc *TransactionContext) UpdateDeltaForTable(physicalTableID int64, delta i
...
 }
 
+// A data struct to store the delta information for a table.
+type ColSize struct {
+	ColID int64
+	Size  int64
+}
+
+// UpdateDeltaForTableFromColSlice is the same as UpdateDeltaForTable, but it accepts a slice of column size.
+func (tc *TransactionContext) UpdateDeltaForTableFromColSlice(
```
Comment (cfzjywxk) [inline pkg/sessionctx/variable/session.go:335]: Should the `UpdateDeltaForTable` be removed after introducing the new method.

Comment (ekexium) [reply pkg/sessionctx/variable/session.go:335]: It's still used in other places, e.g. importer.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Review by zyguan — APPROVED** [review]

**Comment by ekexium** [comment]
@yudongusa PTAL

**Review by you06 — APPROVED** [review]

**Comment by cfzjywxk** [comment]
@ekexium 
Need approvement from variable file owners.

**Review by lcwangchao — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
LGTM
